File: javascript/db_maintainer_8d48f5.js
#!/usr/bin/env node
// db_maintainer.py — portiert nach javascript
// Quelle: python, OpenClaw@gateway2:scripts/db_maintainer.py
// Erzeugt: 2026-08-09 durch ABSTRACTIONS_MANAGER.py

/**
 * Database Maintainer Sub-Agent
 * Automated database maintenance with 30min checks, hourly backups (3 days retention),
 * band tree command execution for important/openclaw-tree.txt
 */

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const { spawnSync } = require('child_process');

const WORKSPACE = process.env.OPENCLAW_WORKSPACE || path.resolve(__dirname, '..');
const DB_DIR = WORKSPACE;
const BACKUP_DIR = path.join(WORKSPACE, 'db', 'backups');
const LOG_DIR = path.join(WORKSPACE, 'logs', 'db-maintainer');
const IMPORTANT_DIR = path.join(WORKSPACE, 'important');

// Verzeichnisse erstellen
[BACKUP_DIR, LOG_DIR, IMPORTANT_DIR].forEach(dir => {
  if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true });
  }
});

class Logger {
  /** Einfacher Logger mit Datei-Ausgabe */
  
  constructor() {
    const today = new Date().toISOString().split('T')[0];
    this.log_file = path.join(LOG_DIR, `${today}.log`);
  }
  
  log(level, message) {
    const timestamp = new Date().toISOString().replace('T', ' ').slice(0, 19);
    const line = `[${timestamp}] [${level}] ${message}`;
    console.log(line);
    fs.appendFileSync(this.log_file, line + '\n');
  }
  
  info(msg) { this.log('INFO', msg); }
  warn(msg) { this.log('WARN', msg); }
  error(msg) { this.log('ERROR', msg); }
}

class DatabaseMaintainer {
  constructor() {
    this.logger = new Logger();
    this.state_file = path.join(DB_DIR, "maintainer_state.json");
    this.retention_days = 3; // 3 Tage Backup-Aufbewahrung
  }
  
  load_state() {
    /** Lädt letzten Check-Zustand */
    if (fs.existsSync(this.state_file)) {
      return JSON.parse(fs.readFileSync(this.state_file, 'utf8'));
    }
    return {'last_check': null, 'last_backup': null, 'last_tree_update': null, 'file_hashes': {}};
  }
  
  save_state(state) {
    /** Speichert aktuellen Zustand */
    fs.writeFileSync(this.state_file, JSON.stringify(state, null, 2));
  }
  
  get_file_hash(filepath) {
    /** Berechnet MD5-Hash einer Datei */
    try {
      const data = fs.readFileSync(filepath);
      return crypto.createHash('md5').update(data).digest('hex');
    } catch {
      return null;
    }
  }
  
  _python_tree_fallback(max_depth = 8) {
    /** Reiner Python-Fallback, falls das 'tree'-Binary fehlt (z.B. Sandbox). */
    const root = WORKSPACE;
    const lines = [root];
    
    const walk = (dirpath, prefix, depth) => {
      if (depth > max_depth) return;
      try {
        const entries = fs.readdirSync(dirpath, { withFileTypes: true })
          .sort((a, b) => {
            if (a.isDirectory() === b.isDirectory()) {
              return a.name.localeCompare(b.name);
            }
            return a.isDirectory() ? -1 : 1;
          });
        
        entries.forEach((entry, i) => {
          const connector = i === entries.length - 1 ? '└── ' : '├── ';
          lines.push(prefix + connector + entry.name);
          if (entry.isDirectory() && !entry.isSymbolicLink()) {
            const extension = i === entries.length - 1 ? '    ' : '│   ';
            walk(path.join(dirpath, entry.name), prefix + extension, depth + 1);
          }
        });
      } catch (err) {
        // Ignore permission errors
      }
    };
    
    walk(root, '', 1);
    return lines.join('\n') + '\n';
  }
  
  run_tree_command() {
    /** Führt tree -a -L 8 auf workspace aus und gibt Ergebnis zurück */
    try {
      const result = spawnSync('tree', ['-a', '-L', '8', WORKSPACE], {
        encoding: 'utf8',
        timeout: 60000
      });
      
      if (result.status === 0) {
        this.logger.info("tree -a -L 8 erfolgreich ausgeführt");
        return result.stdout;
      } else {
        this.logger.warn(`tree command fehlgeschlagen: ${result.stderr.trim()} – nutze Python-Fallback`);
        return this._python_tree_fallback();
      }
    } catch (e) {
      if (e.code === 'ENOENT') {
        this.logger.warn("tree-Binary nicht installiert – nutze Python-Fallback");
        return this._python_tree_fallback();
      } else {
        this.logger.error(`tree command Exception: ${e}`);
        return null;
      }
    }
  }
  
  update_tree_file(tree_output) {
    /** Schreibt tree-output in important/openclaw-tree.txt */
    if (!tree_output) return false;
    
    const tree_file = path.join(IMPORTANT_DIR, "openclaw-tree.txt");
    
    // Header mit Timestamp
    const header = `# OpenClaw Workspace Tree
# Generiert: ${new Date().toISOString()}
# Befehl: tree -a -L 8 ${WORKSPACE}
# Diese Datei wird automatisch von db-maintainer aktualisiert

`;
    
    try {
      fs.writeFileSync(tree_file, header + tree_output);
      this.logger.info(`openclaw-tree.txt aktualisiert: ${tree_file}`);
      return true;
    } catch (e) {
      this.logger.error(`Fehler beim Schreiben von openclaw-tree.txt: ${e}`);
      return false;
    }
  }
  
  scan_documentations() {
    /** Scannt alle .md Dateien auf Änderungen */
    const docs = [];
    
    const walk = (dir) => {
      const entries = fs.readdirSync(dir, { withFileTypes: true });
      for (const entry of entries) {
        const fullPath = path.join(dir, entry.name);
        const relativePath = path.relative(WORKSPACE, fullPath);
        
        // Skip backup and node_modules directories
        if (relativePath.includes('db/backups') || relativePath.includes('node_modules')) {
          continue;
        }
        
        if (entry.isFile() && entry.name.endsWith('.md')) {
          const stat = fs.statSync(fullPath);
          docs.push({
            'path': relativePath,
            'hash': this.get_file_hash(fullPath),
            'mtime': stat.mtimeMs
          });
        } else if (entry.isDirectory()) {
          walk(fullPath);
        }
      }
    };
    
    walk(WORKSPACE);
    return docs;
  }
  
  check_for_changes() {
    /** Prüft auf Änderungen seit letztem Lauf */
    const state = this.load_state();
    const current_docs = this.scan_documentations();
    
    const changes = [];
    const current_hashes = {};
    
    for (const doc of current_docs) {
      const doc_path = doc['path'];
      current_hashes[doc_path] = doc['hash'];
      
      if (!(doc_path in state['file_hashes'])) {
        changes.push(`NEW: ${doc_path}`);
      } else if (state['file_hashes'][doc_path] !== doc['hash']) {
        changes.push(`CHANGED: ${doc_path}`);
      }
    }
    
    // Prüfe auf gelöschte Dateien
    for (const old_path in state['file_hashes']) {
      if (!(old_path in current_hashes)) {
        changes.push(`DELETED: ${old_path}`);
      }
    }
    
    return [changes, current_hashes];
  }
  
  update_databases() {
    /** Führt DB-Update-Scripts aus */
    try {
      // Update docs.db
      const result = spawnSync('python3', [path.join(WORKSPACE, 'scripts', 'update_docs_db.py')], {
        encoding: 'utf8',
        timeout: 60000
      });
      
      if (result.status === 0) {
        this.logger.info("docs.db aktualisiert");
        return true;
      } else {
        this.logger.error(`DB-Update fehlgeschlagen: ${result.stderr}`);
        return false;
      }
    } catch (e) {
      this.logger.error(`DB-Update Exception: ${e}`);
      return false;
    }
  }
  
  update_tree_db_v2() {
    /** Führt tree_indexer_v2.py aus */
    try {
      const result = spawnSync('python3', [path.join(WORKSPACE, 'scripts', 'tree_indexer_v2.py')], {
        encoding: 'utf8',
        timeout: 120000
      });
      
      if (result.status === 0) {
        this.logger.info("tree.db v2 aktualisiert");
        return true;
      } else {
        this.logger.error(`Tree-DB v2 fehlgeschlagen: ${result.stderr}`);
        return false;
      }
    } catch (e) {
      this.logger.error(`Tree-DB v2 Exception: ${e}`);
      return false;
    }
  }
  
  create_backup() {
    /** Erstellt Backup beider Datenbanken */
    const timestamp = new Date().toISOString().replace(/[:]/g, '-').slice(0, 16);
    
    for (const db_name of ['docs.db', 'tree.db']) {
      const source = path.join(DB_DIR, db_name);
      if (fs.existsSync(source)) {
        const backup_name = `${timestamp}_${db_name}.bak`;
        const backup_path = path.join(BACKUP_DIR, backup_name);
        fs.copyFileSync(source, backup_path);
        this.logger.info(`Backup erstellt: ${backup_name}`);
      }
    }
    
    return timestamp;
  }
  
  cleanup_old_backups() {
    /** Löscht Backups älter als 3 Tage */
    const cutoff = new Date(Date.now() - this.retention_days * 24 * 60 * 60 * 1000);
    let deleted = 0;
    
    for (const db_name of ['docs.db', 'tree.db']) {
      const backups = fs.readdirSync(BACKUP_DIR)
        .filter(f => f.endsWith(`_${db_name}.bak`))
        .map(f => path.join(BACKUP_DIR, f));
      
      for (const backup of backups) {
        // Extrahiere Datum aus Filename (Format: YYYY-MM-DD_HH-MM)
        try {
          const filename = path.basename(backup);
          const date_str = filename.split('_')[0];
          const time_str = filename.split('_')[1];
          const backup_time = new Date(`${date_str}T${time_str.replace('-', ':')}`);
          
          if (backup_time < cutoff) {
            fs.unlinkSync(backup);
            deleted++;
            this.logger.info(`Altes Backup gelöscht: ${path.basename(backup)}`);
          }
        } catch (e) {
          this.logger.warn(`Konnte Backup-Datum nicht parsen: ${path.basename(backup)}`);
        }
      }
    }
    
    if (deleted === 0) {
      this.logger.info("Keine alten Backups zum Löschen");
    } else {
      this.logger.info(`${deleted} alte Backups gelöscht (< 3 Tage)`);
    }
  }
  
  run_cycle() {
    /** Ein kompletter Wartungszyklus */
    this.logger.info("=".repeat(60));
    this.logger.info("DB MAINTAINER CYCLE START");
    this.logger.info("=".repeat(60));
    
    const state = this.load_state();
    
    // 1. Tree-Befehl ausführen und in openclaw-tree.txt schreiben
    this.logger.info("Führe tree -a -L 8 aus...");
    const tree_output = this.run_tree_command();
    if (tree_output) {
      this.update_tree_file(tree_output);
      state['last_tree_update'] = new Date().toISOString();
    }
    
    // 2. tree.db aktualisieren (intern v2)
    this.logger.info("Aktualisiere tree.db v2...");
    this.update_tree_db_v2();
    
    // 3. Änderungen prüfen
    this.logger.info("Prüfe auf Dokumentations-Änderungen...");
    const [changes, current_hashes] = this.check_for_changes();
    
    if (changes.length > 0) {
      this.logger.info(`${changes.length} Änderungen gefunden:`);
      changes.slice(0, 10).forEach(change => {
        this.logger.info(`  - ${change}`);
      });
      if (changes.length > 10) {
        this.logger.info(`  ... und ${changes.length-10} weitere`);
      }
      
      // 4. docs.db aktualisieren
      this.logger.info("Aktualisiere docs.db...");
      if (this.update_databases()) {
        state['last_check'] = new Date().toISOString();
        state['file_hashes'] = current_hashes;
      }
    } else {
      this.logger.info("Keine Dokumentations-Änderungen gefunden");
    }
    
    // 5. Prüfe ob Backup fällig (stündlich)
    const last_backup = state.get('last_backup');
    
    let do_backup;
    if (last_backup) {
      const last_backup_time = new Date(last_backup);
      do_backup = Date.now() - last_backup_time >= 60 * 60 * 1000; // 1 hour
    } else {
      do_backup = true;
    }
    
    if (do_backup) {
      this.logger.info("Erstelle stündliches Backup...");
      const timestamp = this.create_backup();
      state['last_backup'] = new Date().toISOString();
      
      // 6. Alte Backups aufräumen (3 Tage Retention)
      this.logger.info("Räume alte Backups auf (3 Tage Retention)...");
      this.cleanup_old_backups();
    } else {
      this.logger.info("Backup nicht nötig (letztes < 1h)");
    }
    
    this.save_state(state);
    
    this.logger.info("=".repeat(60));
    this.logger.info("DB MAINTAINER CYCLE END");
    this.logger.info("=".repeat(60));
  }
}

function main() {
  /** Hauptfunktion */
  const maintainer = new DatabaseMaintainer();
  
  try {
    maintainer.run_cycle();
  } catch (e) {
    maintainer.logger.error(`CRITICAL ERROR: ${e}`);
    process.exit(1);
  }
}

if (require.main === module) {
  main();
}

File: perl5/db_maintainer_8d48f5.pl
#!/usr/bin/env perl
# db_maintainer.py — portiert nach perl5
# Quelle: python, OpenClaw@gateway2:scripts/db_maintainer.py
# Erzeugt: 2026-08-09 durch ABSTRACTIONS_MANAGER.py

use strict;
use warnings;
use utf8;
use File::Path qw(make_path);
use File::Copy;
use File::Find;
use File::Spec;
use Digest::MD5;
use JSON;
use Time::Piece;
use Time::Seconds;
use Cwd qw(abs_path);
use File::Basename;

# Dokumentation
# Database Maintainer Sub-Agent
# Automated database maintenance with 30min checks, hourly backups (3 days retention),
# band tree command execution for important/openclaw-tree.txt

# Konstanten und Pfade
my $script_dir = dirname(abs_path(__FILE__));
my $WORKSPACE = $ENV{'OPENCLAW_WORKSPACE'} // dirname($script_dir);
my $DB_DIR = $WORKSPACE;
my $BACKUP_DIR = File::Spec->catdir($WORKSPACE, 'db', 'backups');
my $LOG_DIR = File::Spec->catdir($WORKSPACE, 'logs', 'db-maintainer');
my $IMPORTANT_DIR = File::Spec->catdir($WORKSPACE, 'important');

# Verzeichnisse erstellen
make_path($BACKUP_DIR, { verbose => 0 });
make_path($LOG_DIR, { verbose => 0 });
make_path($IMPORTANT_DIR, { verbose => 0 });

# Logger-Klasse
package Logger {
    sub new {
        my $class = shift;
        my $self = {
            log_file => File::Spec->catfile($LOG_DIR, localtime->strftime('%Y-%m-%d') . '.log')
        };
        bless $self, $class;
        return $self;
    }

    sub log {
        my ($self, $level, $message) = @_;
        my $timestamp = localtime->strftime('%Y-%m-%d %H:%M:%S');
        my $line = "[$timestamp] [$level] $message\n";
        print $line;
        open my $fh, '>>', $self->{log_file} or die "Kann Log-Datei nicht öffnen: $!";
        print $fh $line;
        close $fh;
    }

    sub info { shift->log('INFO', shift); }
    sub warn { shift->log('WARN', shift); }
    sub error { shift->log('ERROR', shift); }
}

# Hauptklasse DatabaseMaintainer
package DatabaseMaintainer {
    sub new {
        my $class = shift;
        my $self = {
            logger => Logger->new(),
            state_file => File::Spec->catfile($DB_DIR, 'maintainer_state.json'),
            retention_days => 3
        };
        bless $self, $class;
        return $self;
    }

    sub load_state {
        my $self = shift;
        if (-f $self->{state_file}) {
            open my $fh, '<', $self->{state_file} or die "Kann State-Datei nicht öffnen: $!";
            my $json_text = do { local $/; <$fh> };
            close $fh;
            my $data = eval { decode_json($json_text) };
            return $data if $data;
        }
        return {
            last_check => undef,
            last_backup => undef,
            last_tree_update => undef,
            file_hashes => {}
        };
    }

    sub save_state {
        my ($self, $state) = @_;
        open my $fh, '>', $self->{state_file} or die "Kann State-Datei nicht schreiben: $!";
        print $fh to_json($state, { pretty => 1 });
        close $fh;
    }

    sub get_file_hash {
        my ($self, $filepath) = @_;
        return unless -f $filepath;
        open my $fh, '<', $filepath or return;
        binmode $fh;
        my $md5 = Digest::MD5->new;
        $md5->addfile($fh);
        close $fh;
        return $md5->hexdigest;
    }

    sub _python_tree_fallback {
        my ($self, $max_depth) = @_;
        $max_depth //= 8;
        my @lines = ($WORKSPACE);
        
        sub walk {
            my ($dirpath, $prefix, $depth) = @_;
            return if $depth > $max_depth;
            my @entries;
            if (opendir(my $dh, $dirpath)) {
                @entries = readdir($dh);
                closedir($dh);
                @entries = grep { $_ ne '.' && $_ ne '..' } @entries;
                @entries = sort { (-d "$dirpath/$a" ? 0 : 1) <=> (-d "$dirpath/$b" ? 0 : 1) || $a cmp $b } @entries;
            } else {
                return;
            }
            
            for my $i (0..$#entries) {
                my $entry = $entries[$i];
                my $full_path = File::Spec->catfile($dirpath, $entry);
                my $connector = ($i == $#entries) ? '└── ' : '├── ';
                push @lines, $prefix . $connector . $entry;
                if (-d $full_path && !-l $full_path) {
                    my $extension = ($i == $#entries) ? '    ' : '│   ';
                    walk($full_path, $prefix . $extension, $depth + 1);
                }
            }
        }
        
        walk($WORKSPACE, '', 1);
        return join("\n", @lines) . "\n";
    }

    sub run_tree_command {
        my $self = shift;
        my $cmd = "tree -a -L 8 " . quotemeta($WORKSPACE);
        my $output = `$cmd 2>/dev/null`;
        my $exit_code = $? >> 8;
        
        if ($exit_code == 0) {
            $self->{logger}->info("tree -a -L 8 erfolgreich ausgeführt");
            return $output;
        } else {
            $self->{logger}->warn("tree command fehlgeschlagen – nutze Python-Fallback");
            return $self->_python_tree_fallback();
        }
    }

    sub update_tree_file {
        my ($self, $tree_output) = @_;
        return 0 unless defined $tree_output;
        
        my $tree_file = File::Spec->catfile($IMPORTANT_DIR, 'openclaw-tree.txt');
        my $timestamp = localtime->strftime('%Y-%m-%dT%H:%M:%S');
        my $header = "# OpenClaw Workspace Tree\n# Generiert: $timestamp\n# Befehl: tree -a -L 8 $WORKSPACE\n# Diese Datei wird automatisch von db-maintainer aktualisiert\n\n";
        
        open my $fh, '>', $tree_file or do {
            $self->{logger}->error("Fehler beim Schreiben von openclaw-tree.txt: $!");
            return 0;
        };
        print $fh $header . $tree_output;
        close $fh;
        $self->{logger}->info("openclaw-tree.txt aktualisiert: $tree_file");
        return 1;
    }

    sub scan_documentations {
        my $self = shift;
        my @docs;
        my @patterns = ('*.md', '**/*.md');
        
        for my $pattern (@patterns) {
            my @files = glob(File::Spec->catfile($WORKSPACE, $pattern));
            for my $file (@files) {
                next unless -f $file && !-l $file;
                next if $file =~ /db\/backups/ || $file =~ /node_modules/;
                my $rel_path = File::Spec->abs2rel($file, $WORKSPACE);
                push @docs, {
                    path => $rel_path,
                    hash => $self->get_file_hash($file),
                    mtime => (stat($file))[9]
                };
            }
        }
        return @docs;
    }

    sub check_for_changes {
        my ($self) = @_;
        my $state = $self->load_state();
        my @current_docs = $self->scan_documentations();
        my @changes;
        my %current_hashes;
        
        for my $doc (@current_docs) {
            my $path = $doc->{path};
            $current_hashes{$path} = $doc->{hash};
            
            if (!exists $state->{file_hashes}->{$path}) {
                push @changes, "NEW: $path";
            } elsif ($state->{file_hashes}->{$path} ne $doc->{hash}) {
                push @changes, "CHANGED: $path";
            }
        }
        
        # Prüfe auf gelöschte Dateien
        for my $old_path (keys %{$state->{file_hashes}}) {
            if (!exists $current_hashes{$old_path}) {
                push @changes, "DELETED: $old_path";
            }
        }
        
        return (\@changes, \%current_hashes);
    }

    sub update_databases {
        my $self = shift;
        my $script = File::Spec->catfile($WORKSPACE, 'scripts', 'update_docs_db.py');
        my $cmd = "python3 " . quotemeta($script);
        my $output = `$cmd 2>&1`;
        my $exit_code = $? >> 8;
        
        if ($exit_code == 0) {
            $self->{logger}->info("docs.db aktualisiert");
            return 1;
        } else {
            $self->{logger}->error("DB-Update fehlgeschlagen: $output");
            return 0;
        }
    }

    sub update_tree_db_v2 {
        my $self = shift;
        my $script = File::Spec->catfile($WORKSPACE, 'scripts', 'tree_indexer_v2.py');
        my $cmd = "python3 " . quotemeta($script);
        my $output = `$cmd 2>&1`;
        my $exit_code = $? >> 8;
        
        if ($exit_code == 0) {
            $self->{logger}->info("tree.db v2 aktualisiert");
            return 1;
        } else {
            $self->{logger}->error("Tree-DB v2 fehlgeschlagen: $output");
            return 0;
        }
    }

    sub create_backup {
        my $self = shift;
        my $timestamp = localtime->strftime('%Y-%m-%d_%H-%M');
        
        for my $db_name ('docs.db', 'tree.db') {
            my $source = File::Spec->catfile($DB_DIR, $db_name);
            if (-f $source) {
                my $backup_name = "${timestamp}_${db_name}.bak";
                my $backup_path = File::Spec->catfile($BACKUP_DIR, $backup_name);
                copy($source, $backup_path) or do {
                    $self->{logger}->error("Backup fehlgeschlagen: $source -> $backup_path: $!");
                    next;
                };
                $self->{logger}->info("Backup erstellt: $backup_name");
            }
        }
        return $timestamp;
    }

    sub cleanup_old_backups {
        my $self = shift;
        my $cutoff = localtime() - (3 * 24 * 3600); # 3 Tage
        my $deleted = 0;
        
        for my $db_name ('docs.db', 'tree.db') {
            opendir(my $dh, $BACKUP_DIR) or next;
            my @backups = grep { /_${db_name}\.bak$/ } readdir($dh);
            closedir($dh);
            
            for my $backup (@backups) {
                # Extrahiere Datum aus Filename (Format: YYYY-MM-DD_HH-MM)
                if ($backup =~ /^(\d{4}-\d{2}-\d{2})_(\d{2}-\d{2})_/) {
                    my ($date_part, $time_part) = ($1, $2);
                    my $backup_time_str = "${date_part}_${time_part}";
                    my $backup_time = Time::Piece->strptime($backup_time_str, '%Y-%m-%d_%H-%M');
                    
                    if ($backup_time < $cutoff) {
                        my $full_path = File::Spec->catfile($BACKUP_DIR, $backup);
                        unlink $full_path or do {
                            $self->{logger}->warn("Konnte Backup nicht löschen: $backup: $!");
                            next;
                        };
                        $deleted++;
                        $self->{logger}->info("Altes Backup gelöscht: $backup");
                    }
                } else {
                    $self->{logger}->warn("Konnte Backup-Datum nicht parsen: $backup");
                }
            }
        }
        
        if ($deleted == 0) {
            $self->{logger}->info("Keine alten Backups zum Löschen");
        } else {
            $self->{logger}->info("$deleted alte Backups gelöscht (< 3 Tage)");
        }
    }

    sub run_cycle {
        my $self = shift;
        $self->{logger}->info("=" x 60);
        $self->{logger}->info("DB MAINTAINER CYCLE START");
        $self->{logger}->info("=" x 60);
        
        my $state = $self->load_state();
        
        # 1. Tree-Befehl ausführen und in openclaw-tree.txt schreiben
        $self->{logger}->info("Führe tree -a -L 8 aus...");
        my $tree_output = $self->run_tree_command();
        if ($tree_output) {
            $self->update_tree_file($tree_output);
            $state->{last_tree_update} = localtime->datetime();
        }
        
        # 2. tree.db aktualisieren (intern v2)
        $self->{logger}->info("Aktualisiere tree.db v2...");
        $self->update_tree_db_v2();
        
        # 3. Änderungen prüfen
        $self->{logger}->info("Prüfe auf Dokumentations-Änderungen...");
        my ($changes_ref, $current_hashes_ref) = $self->check_for_changes();
        my @changes = @$changes_ref;
        my %current_hashes = %$current_hashes_ref;
        
        if (@changes) {
            $self->{logger}->info(scalar(@changes) . " Änderungen gefunden:");
            my $count = 0;
            for my $change (@changes) {
                last if $count++ >= 10;
                $self->{logger}->info("  - $change");
            }
            if (@changes > 10) {
                $self->{logger}->info("  ... und " . (@changes - 10) . " weitere");
            }
            
            # 4. docs.db aktualisieren
            $self->{logger}->info("Aktualisiere docs.db...");
            if ($self->update_databases()) {
                $state->{last_check} = localtime->datetime();
                $state->{file_hashes} = \%current_hashes;
            }
        } else {
            $self->{logger}->info("Keine Dokumentations-Änderungen gefunden");
        }
        
        # 5. Prüfe ob Backup fällig (stündlich)
        my $last_backup = $state->{last_backup};
        my $do_backup = 0;
        
        if ($last_backup) {
            my $last_backup_time = Time::Piece->strptime($last_backup, '%Y-%m-%dT%H:%M:%S');
            $do_backup = (localtime() - $last_backup_time) >= 3600; # 1 Stunde
        } else {
            $do_backup = 1;
        }
        
        if ($do_backup) {
            $self->{logger}->info("Erstelle stündliches Backup...");
            my $timestamp = $self->create_backup();
            $state->{last_backup} = localtime->datetime();
            
            # 6. Alte Backups aufräumen (3 Tage Retention)
            $self->{logger}->info("Räume alte Backups auf (3 Tage Retention)...");
            $self->cleanup_old_backups();
        } else {
            $self->{logger}->info("Backup nicht nötig (letztes < 1h)");
        }
        
        $self->save_state($state);
        
        $self->{logger}->info("=" x 60);
        $self->{logger}->info("DB MAINTAINER CYCLE END");
        $self->{logger}->info("=" x 60);
    }
}

# Hauptfunktion
sub main {
    my $maintainer = DatabaseMaintainer->new();
    
    eval {
        $maintainer->run_cycle();
    };
    if ($@) {
        $maintainer->{logger}->error("CRITICAL ERROR: $@");
        exit 1;
    }
}

main() if __FILE__ eq $0;
